src/core/border_router/routing_manager.cpp

```diff
@@ -334,40 +358,43 @@ void RoutingManager::HandleReceived(const InfraIf::Icmp6Packet &aPacket, const I
 
 void RoutingManager::HandleNotifierEvents(Events aEvents)
 {
-    VerifyOrExit(IsInitialized() && IsEnabled());
-
-    if (aEvents.Contains(kEventThreadRoleChanged))
+    switch (GetState())
     {
-        EvaluateState();
-    }
+    case kStateUninitialized:
+    case kStateDisabled:
+        break;
 
-    if (mIsRunning && aEvents.Contains(kEventThreadNetdataChanged))
-    {
-        // Invalidate discovered prefixes because OMR Prefixes in Network Data may change.
-        InvalidateDiscoveredPrefixes();
-        StartRoutingPolicyEvaluationJitter(kRoutingPolicyEvaluationJitter);
-    }
+    case kStateEnabledRunning:
+        if (aEvents.Contains(kEventThreadNetdataChanged))
+        {
+            // Invalidate discovered prefixes because OMR Prefixes in Network Data may change.
+            InvalidateDiscoveredPrefixes();
+            StartRoutingPolicyEvaluationJitter(kRoutingPolicyEvaluationJitter);
+        }
 
-    if (aEvents.Contains(kEventThreadExtPanIdChanged))
-    {
-        if (mIsAdvertisingLocalOnLinkPrefix)
+        if (aEvents.Contains(kEventThreadExtPanIdChanged))
         {
-            UnpublishExternalRoute(mLocalOnLinkPrefix);
-            // TODO: consider deprecating/invalidating existing
-            // on-link prefix
-            mIsAdvertisingLocalOnLinkPrefix = false;
+            if (mIsAdvertisingLocalOnLinkPrefix)
+            {
+                UnpublishExternalRoute(mLocalOnLinkPrefix);
+                // TODO: consider deprecating/invalidating existing
+                // on-link prefix
+                mIsAdvertisingLocalOnLinkPrefix = false;
+            }
+
+            GenerateOnLinkPrefix();
+            StartRoutingPolicyEvaluationJitter(kRoutingPolicyEvaluationJitter);
         }
 
-        GenerateOnLinkPrefix();
+        OT_FALL_THROUGH;
 
-        if (mIsRunning)
+    case kStateEnabledIdle:
+        if (aEvents.Contains(kEventThreadRoleChanged))
         {
-            StartRoutingPolicyEvaluationJitter(kRoutingPolicyEvaluationJitter);
+            EvaluateState();
         }
+        break;
```

Comment:
Can we receive both `kEventThreadRoleChanged` and `kEventThreadNetdataChanged` events when we are in the `kStateEnabledIdle` state?

Reply:
We only process `kEventThreadNetdataChanged` under `mIsRunning` before which now maps to `kStateEnabledRunning` (we can ignore it otherwise).